pose_prediction/plugin.cpp

```diff
@@ -14,33 +14,40 @@ class pose_prediction_impl : public pose_prediction {
 		, _m_slow_pose{sb->subscribe_latest<pose_type>("slow_pose")}
         , _m_imu_raw{sb->subscribe_latest<imu_raw_type>("imu_raw")}
         , _m_true_pose{sb->subscribe_latest<pose_type>("true_pose")}
+        , _m_ground_truth_offset{sb->subscribe_latest<Eigen::Vector3f>("ground_truth_offset")}
         , _m_vsync_estimate{sb->subscribe_latest<time_type>("vsync_estimate")}
     { }
 
-    // No paramter get_fast_pose() should just predict to the next vsync
-	// However, we don't have vsync estimation yet.
-	// So we will predict to `now()`, as a temporary approximation
+    // No parameter get_fast_pose() predicts to the next vsync if a vsync
+    // estimate is available, and predicts to `now` otherwise
     virtual fast_pose_type get_fast_pose() const override {
 		const time_type *vsync_estimate = _m_vsync_estimate->get_latest_ro();
 
-        if(vsync_estimate == nullptr) {
-		return get_fast_pose(std::chrono::high_resolution_clock::now());
+        if (vsync_estimate == nullptr) {
+            return get_fast_pose(std::chrono::high_resolution_clock::now());
         } else {
             return get_fast_pose(*vsync_estimate);
         }
-	}
-
-    virtual pose_type get_true_pose() const override {
-		const pose_type* pose_ptr = _m_true_pose->get_latest_ro();
-		return correct_pose(
-			pose_ptr ? *pose_ptr : pose_type{
-				.sensor_time = std::chrono::system_clock::now(),
-				.position = Eigen::Vector3f{0, 0, 0},
-				.orientation = Eigen::Quaternionf{1, 0, 0, 0},
-			}
-		);
     }
 
+	virtual pose_type get_true_pose() const override {
+		const auto *pose = _m_true_pose->get_latest_ro();
+		const auto *offset = _m_ground_truth_offset->get_latest_ro();
+		pose_type offset_pose;
+
+		// Subtract offset if valid pose and offset, otherwise use zero pose
+		if (pose && offset) {
```

Comment:
Can `(pose && !offset) || (!pose && offset)` be true? If so, when?

Reply:
Hypothetically, if switchboard moved to a weaker memory model, writes to two different streams could finish out of order, resulting in one stream being valid and the other being invalid.

Checking that both are valid is safer because it assumes nothing about the ordering of writes on the producer's end or about the producer actually writing to both.

Reply:
Awesome! Can we add that in the comments? :smiley:

Reply:
Hah, will do.